### PR TITLE
Enhance release gate validation and expand test coverage

### PR DESCRIPTION
**Pull request by cnkang**
## Summary by Sourcery

Expand NGINX coverage tooling and tests while tightening release gate validation and documenting coverage standards.

New Features:
- Add comprehensive NGINX coverage nginx.conf and curl scenarios to exercise auth detection, conditional requests, error handling, Accept header parsing, metrics formats, and markdown flavors.
- Introduce advisory per-file and aggregate coverage threshold checks in the NGINX coverage collection script.
- Add a dedicated unit test suite for auth cookie pattern matching covering wildcard and edge cases.

Enhancements:
- Extend Accept header parser unit tests with q-value edge cases, malformed inputs, and range invariants to strengthen behavior guarantees.
- Expand error classification unit tests to cover success mapping, unknown categories, and completeness of error-to-category mappings.
- Make spec validation tolerant of archived sub-spec directories when checking for required requirements/design documents and their sections.
- Clarify coverage standards, expectations, and workflows in harness, testing, and agent documentation, including when to run coverage commands and target thresholds.

Documentation:
- Document coverage E2E scenarios and coverage standards in testing and harness docs, including commands and target thresholds for C and Rust components.
- Update agent workflow guidelines to require coverage checks and maintain per-file and aggregate coverage bars for production changes.

Tests:
- Add new auth cookie pattern unit tests and broaden Accept parser, error classification, and reason code tests to improve C module coverage.

## Release Notes

* **Tests**
  * Added comprehensive unit tests for Accept header parsing, authentication patterns, and error handling
  * Expanded E2E test scenarios covering authentication flows, conditional requests, error paths, and metrics collection

* **Documentation**
  * Established testing coverage standards (80% minimum aggregate, 90% for critical code paths)
  * Updated testing documentation with expanded scenario coverage and collection procedures

### CODE REVIEW

File: tools/sonar/collect_nginx_coverage.sh [hunk 1]
```diff
@@ -249,30 +338,137 @@ sleep 1
...
-curl -sS -I -H 'Accept: text/markdown' "http://127.0.0.1:${PORT}/index.html" -o /dev/null -w "  HEAD request: HTTP %{http_code}\n"
+curl -sS -I -H "${ACCEPT_MARKDOWN}" "http://127.0.0.1:${PORT}/index.html" -o /dev/null -w "  HEAD request: HTTP %{http_code}\n"
+
+# Large page (chain accumulation)
+curl -sS -H "${ACCEPT_MARKDOWN}" "http://127.0.0.1:${PORT}/large.html" -o /dev/null -w "  large page: HTTP %{http_code}\n"
+
+# ── Auth detection scenarios (Req 2) ────────────────────────────────
+
+# Bearer token auth (dummy placeholder — exercises auth detection code path)
+curl -sS -H "${ACCEPT_MARKDOWN}" -H 'Authorization: Bearer DUMMY_TEST_TOKEN' \
```
[the remark this answers is not in the document]
Comment (cnkang) [reply tools/sonar/collect_nginx_coverage.sh:361]: DUMMY_TEST_TOKEN, false postive

File: components/nginx-module/tests/unit/accept_parser_test.c [hunk 2]
```diff
@@ -270,6 +270,115 @@ test_malformed_entries(void)
...
+static void
+test_q_value_edge_cases(void)
+{
+    TEST_SUBSECTION("Q-value edge cases");
+
+    /* q=1 (maximum valid) */
+    TEST_ASSERT(should_convert("text/markdown;q=1", 0) == 1,
+                "q=1 should convert");
+
+    /* q=1.000 (maximum valid with decimals) */
```
[the remark this answers is not in the document]
Comment (cnkang) [reply components/nginx-module/tests/unit/accept_parser_test.c:284]: Fixed in commit af64029. Added tests for `q=-0.1` (clamped to 0.0), `q= 0.5` (leading space), `q=0.5 ` (trailing space), and `q =0.5` (space before '=').

File: components/nginx-module/tests/unit/error_classification_test.c [hunk 3]
```diff
@@ -80,6 +80,58 @@ test_system_errors_and_strings(void)
...
+static void
+test_error_success_mapping(void)
+{
+    TEST_SUBSECTION("ERROR_SUCCESS (code 0) maps to system default");
+    TEST_ASSERT(classify_error(ERROR_SUCCESS) == CAT_SYSTEM,
+                "ERROR_SUCCESS (0) should map to CAT_SYSTEM (default)");
+    TEST_PASS("ERROR_SUCCESS mapping correct");
+}
+
+static void
```
[the remark this answers is not in the document]
Comment (cnkang) [reply components/nginx-module/tests/unit/error_classification_test.c:95]: Fixed in commit af64029. Added `UINT_MAX` boundary-value test with `#include <limits.h>`.

File: components/nginx-module/tests/unit/auth_cookie_pattern_test.c [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+static void
+test_prefix_wildcard_matching(void)
+{
+    TEST_SUBSECTION("Prefix wildcard matching");
+
+    /* "session*" matches "session_id" */
+    TEST_ASSERT(cookie_matches_pattern("session_id", "session*") == 1,
+                "session* matches session_id");
+    TEST_ASSERT(cookie_matches_pattern("session", "session*") == 1,
+                "session* matches session (exact prefix)");
```
[the remark this answers is not in the document]
Comment (cnkang) [reply components/nginx-module/tests/unit/auth_cookie_pattern_test.c:91]: Fixed in commit af64029. Added tests for `*session*`, `sess*id`, and `**` patterns documenting that they are NOT treated as substring wildcards. Also added DIVERGENCE RISK documentation in the function comment.

File: AGENTS.md [hunk 5]
```diff
@@ -701,6 +701,19 @@ Required:
...
+
+Required:
+- New C module code must include e2e scenarios or unit tests that maintain the
+  80% aggregate line coverage bar.
+- New Rust converter code must include tests that maintain the 80% aggregate
+  line coverage bar.
+- Critical paths (auth, error handling, FFI boundary, conditional requests)
+  require 90% line coverage for new code.
+- Before declaring a task complete when the change touches production source
+  files, run `make coverage-c` (C module) or `make coverage-rust` (Rust
```
[the remark this answers is not in the document]
Comment (cnkang) [reply AGENTS.md:714]: Fixed in commit 7e72152. Rule 25 now explicitly states: "The advisory per-file thresholds logged by collect_nginx_coverage.sh are informational warnings only — the binding gate is the 80% aggregate bar (90% for critical paths)." Checklist items 19 (Rust) and 20 (C) both say "80% (90% for critical paths)" which is consistent.

File: tools/sonar/collect_nginx_coverage.sh [hunk 6]
```diff
@@ -258,22 +346,127 @@ curl -sS -H 'Accept: text/html' "http://127.0.0.1:${PORT}/index.html" -o /dev/nu
...
-# Large page
+# Large page (chain accumulation)
 curl -sS -H 'Accept: text/markdown' "http://127.0.0.1:${PORT}/large.html" -o /dev/null -w "  large page: HTTP %{http_code}\n"
 
-# Metrics endpoint
-curl -sS "http://127.0.0.1:${PORT}/metrics" -o /dev/null -w "  metrics: HTTP %{http_code}\n"
+# ── Auth detection scenarios (Req 2) ────────────────────────────────
+
+# Bearer token auth
+curl -sS -H 'Accept: text/markdown' -H 'Authorization: Bearer token123' \
```
[the remark this answers is not in the document]
Comment (cnkang) [reply tools/sonar/collect_nginx_coverage.sh]: False positive. The token was already changed to `DUMMY_TEST_TOKEN` in a previous commit. This is a test-only dummy value exercising the auth detection code path against a local NGINX instance — no real credentials.

File: tools/sonar/collect_nginx_coverage.sh [hunk 7]
```diff
@@ -310,3 +503,56 @@ echo "==> C module e2e coverage report: ${OUTPUT_LCOV}"
...
+aggregate="$(lcov --summary "${OUTPUT_LCOV}" --rc branch_coverage=1 \
+  ${LCOV_IGNORE} 2>&1 | awk '/lines\.\.\.\./{print $2}' || true)"
+if [[ -n "${aggregate}" ]]; then
+  aggregate_int="${aggregate%%.*}"
+  if [[ -n "${aggregate_int}" ]] && [[ "${aggregate_int}" -lt 80 ]]; then
+    echo "  WARNING: Aggregate line coverage ${aggregate}% below 80% minimum" >&2
+  else
+    echo "  Aggregate line coverage: ${aggregate}% (meets 80% minimum)"
+  fi
+fi
```
[the remark this answers is not in the document]
Comment (cnkang) [reply tools/sonar/collect_nginx_coverage.sh:566]: Valid point. Fixed in commit ee40e5e: softened "release-readiness quality gate" to "release-readiness advisory bar (logged, not CI-blocking)" in core.md to match the actual script behavior (warns only, exits 0). The aggregate 80% is a human-enforced standard during release review, not a CI-blocking gate.

File: tools/sonar/collect_nginx_coverage.sh [hunk 8]
```diff
@@ -310,3 +503,56 @@ echo "==> C module e2e coverage report: ${OUTPUT_LCOV}"
...
+    ngx_http_markdown_accept.c)              threshold=70 ;;
+    ngx_http_markdown_config_handlers_impl.h) threshold=40 ;;
+    *) ;;
+  esac
+
+  if [[ -n "${threshold}" ]] && [[ -n "${pct}" ]]; then
+    # Compare as integers (truncate decimal)
+    pct_int="${pct%%.*}"
+    if [[ -n "${pct_int}" ]] && [[ "${pct_int}" -lt "${threshold}" ]]; then
+      echo "  WARNING: ${file} line coverage ${pct}% below advisory threshold ${threshold}%" >&2
```
[the remark this answers is not in the document]
Comment (cnkang) [reply tools/sonar/collect_nginx_coverage.sh]: Valid bug. Fixed in commit ee40e5e: replaced `awk '{print $NF}'` with pipe-delimited column parsing (`awk -F'|'`) to extract the Lines rate percentage from the correct column, instead of accidentally picking up the branch count from the last whitespace-delimited field.

File: AGENTS.md [hunk 9]
```diff
@@ -770,6 +784,7 @@ For each code change you are about to produce, mentally (or explicitly in a thin
...
 17. **Metric naming accuracy**: when a metric field name implies specific semantics (for example `cpu_time_ms`), the implementation must actually measure that quantity. If approximating with a different measurement (for example wall-clock TTLB), add a comment documenting the approximation or rename the field. (Rule 8)
 18. **Rust 2024 edition pattern binding**: do not use explicit `ref` or `ref mut` binding modifiers when matching on a reference type (`&T`, `&Option<T>`, `&Vec<T>`, etc.). Rust 2024 edition treats this as an error because the reference already implies borrowing. Use `ref` only when matching on an owned value where you need to borrow without moving. Before writing `if let Some(ref x) = expr`, check whether `expr` is a reference — if so, drop the `ref`.
+19. Coverage: if this change adds or modifies production code, verify that `cargo llvm-cov` coverage for the touched files does not regress below 80%. (Rule 25)
```
[the remark this answers is not in the document]
Comment (cnkang) [reply AGENTS.md]: No change needed. The current text already says `80% (90% for critical paths)` which is exactly consistent with Rule 25. Rule 25 explicitly states: "the binding gate is the 80% aggregate bar (90% for critical paths)." The 80% is the general minimum; 90% only applies to critical paths (auth, error handling, FFI boundary, conditional requests). Changing the checklist to 90% across the board would be incorrect.

File: components/nginx-module/tests/unit/auth_cookie_pattern_test.c [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+                        pattern + 1, suffix_len) == 0;
+    }
+
+    /* Exact match */
+    if (cookie_len != pattern_len) {
+        return 0;
+    }
+
+    return strncmp(cookie_name, pattern, pattern_len) == 0;
+}
```
[the remark this answers is not in the document]
Comment (cnkang) [reply components/nginx-module/tests/unit/auth_cookie_pattern_test.c:78]: No change needed. The production function `ngx_http_markdown_cookie_matches_pattern` is `static` and uses NGINX types (`ngx_str_t`, `ngx_int_t`) — it cannot be called from standalone unit tests without pulling in the entire NGINX header chain. The test already documents this constraint with a DIVERGENCE RISK comment. The real production code is exercised by E2E tests in `collect_nginx_coverage.sh`.

File: tools/release/gates/validate_release_gates.py [hunk 11]
```diff
@@ -203,20 +203,37 @@ def _parse_table_cells_positional(line: str) -> list[str]:
...
+    This helper checks the top-level location first, then the archive.
+
+    Returns:
+        The resolved directory ``Path``, or ``None`` if not found.
+    """
+    candidate = specs_dir / name
+    if candidate.is_dir():
+        return candidate
+    archived = specs_dir / "archive" / name
+    return archived if archived.is_dir() else None
```
[the remark this answers is not in the document]
Comment (cnkang) [reply tools/release/gates/validate_release_gates.py]: Fixed in commit 58047d6 (initial hardening with `str.startswith()`) and then improved in commit ee40e5e to use `is_relative_to()` per the greptile comment below, which correctly handles the sibling-prefix bypass case.

File: tools/sonar/collect_nginx_coverage.sh [hunk 12]
```diff
@@ -310,3 +503,56 @@ echo "==> C module e2e coverage report: ${OUTPUT_LCOV}"
...
+    fi
+  fi
```
[the remark this answers is not in the document]
Comment (cnkang) [reply tools/sonar/collect_nginx_coverage.sh:552]: Fixed in commit 58047d6. Added `[[ "${pct_int}" =~ ^[0-9]+$ ]]` numeric validation before the integer comparison to prevent shell syntax errors on non-numeric lcov output.

File: tools/sonar/collect_nginx_coverage.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (cnkang) [reply tools/sonar/collect_nginx_coverage.sh:361]: Duplicate of comment #5. Already fixed — token changed to `DUMMY_TEST_TOKEN` in a previous commit. False positive: this is a test-only dummy value for a local NGINX instance.

File: tools/sonar/collect_nginx_coverage.sh [hunk 13]
```diff
@@ -249,30 +338,137 @@ sleep 1
...
+curl -sS -H "${ACCEPT_MARKDOWN}" -H 'Authorization: Bearer DUMMY_TEST_TOKEN' \
+  "http://127.0.0.1:${PORT}/auth/index.html" -o /dev/null -w "  auth bearer: HTTP %{http_code}\n"
+
+# Cookie prefix match (session*)
+curl -sS -H "${ACCEPT_MARKDOWN}" -H 'Cookie: session_id=abc123' \
+  "http://127.0.0.1:${PORT}/auth/index.html" -o /dev/null -w "  auth cookie prefix: HTTP %{http_code}\n"
+
+# Cookie suffix match (*_logged_in)
+curl -sS -H "${ACCEPT_MARKDOWN}" -H 'Cookie: wordpress_logged_in_hash=val' \
+  "http://127.0.0.1:${PORT}/auth/index.html" -o /dev/null -w "  auth cookie suffix: HTTP %{http_code}\n"
```
[the remark this answers is not in the document]
Comment (cnkang) [reply tools/sonar/collect_nginx_coverage.sh]: Good catch! Fixed in commit ee40e5e. Changed `wordpress_logged_in_hash` to `wordpress_logged_in` so the cookie name actually matches the `*_logged_in` suffix pattern and exercises the suffix-match branch.

File: tools/release/gates/validate_release_gates.py [hunk 14]
```diff
@@ -203,20 +203,67 @@ def _parse_table_cells_positional(line: str) -> list[str]:
...
+
+    specs_resolved = specs_dir.resolve()
+
+    candidate = (specs_dir / name).resolve()
+    if candidate.is_dir() and str(candidate).startswith(str(specs_resolved)):
+        return candidate
+
+    archived = (specs_dir / "archive" / name).resolve()
+    if archived.is_dir() and str(archived).startswith(str(specs_resolved)):
+        return archived
```
[the remark this answers is not in the document]
Comment (cnkang) [reply tools/release/gates/validate_release_gates.py:237]: Good catch on the sibling-prefix bypass. Fixed in commit ee40e5e: replaced `str(candidate).startswith(str(specs_resolved))` with `candidate.is_relative_to(specs_resolved)` (available since Python 3.9, which is our minimum). This correctly handles both the symlink and adjacent-directory false positive cases.
